validator/client/validator.go

```diff
@@ -1212,44 +1209,66 @@ func (v *validator) ChangeHost() {
 func (v *validator) filterAndCacheActiveKeys(ctx context.Context, pubkeys [][fieldparams.BLSPubkeyLength]byte, slot primitives.Slot) ([][fieldparams.BLSPubkeyLength]byte, error) {
 	ctx, span := trace.StartSpan(ctx, "validator.filterAndCacheActiveKeys")
 	defer span.End()
-
+	isEpochStart := slots.IsEpochStart(slot)
+	validatorStatusUnpopulated := false
 	filteredKeys := make([][fieldparams.BLSPubkeyLength]byte, 0)
 	statusRequestKeys := make([][]byte, 0)
+	validatorStatuses := make([]*validatorStatus, 0)
 	for _, k := range pubkeys {
-		_, ok := v.pubkeyToValidatorIndex[k]
-		// Get validator index from RPC server if not found.
+		sta, ok := v.pubkeyToStatus[k]
 		if !ok {
-			i, ok, err := v.validatorIndex(ctx, k)
-			if err != nil {
-				return nil, err
-			}
-			if !ok { // Nothing we can do if RPC server doesn't have validator index.
-				continue
-			}
-			v.pubkeyToValidatorIndex[k] = i
+			validatorStatusUnpopulated = true
+			// skip adding if not found
+			continue
 		}
 		copiedk := k
 		statusRequestKeys = append(statusRequestKeys, copiedk[:])
+		copiedStatus := sta
+		validatorStatuses = append(validatorStatuses, copiedStatus)
 	}
-	resp, err := v.validatorClient.MultipleValidatorStatus(ctx, &ethpb.MultipleValidatorStatusRequest{
-		PublicKeys: statusRequestKeys,
-	})
-	if err != nil {
-		return nil, err
+	if isEpochStart || validatorStatusUnpopulated {
+		resp, err := v.validatorClient.MultipleValidatorStatus(ctx, &ethpb.MultipleValidatorStatusRequest{
+			PublicKeys: statusRequestKeys,
```

Comment:
If the key didn't have a cached status, then it won't be in the `statusRequestKeys`, even though that key may have triggered this code path.

Reply:
shoot yes, on epoch start i should skip the continue and the requests for all keys should be done anyways

Reply:
been a few days I'm going to double check but I think the cache is populated already but been a few days so going to refresh myself and clean this up anyways

Reply:
definitely doing something wrong here will fix

Reply:
should not have had the continue on it...